test/src/test/java/hudson/bugs/JnlpAccessWithSecuredHudsonTest.java

```diff
@@ -92,47 +106,43 @@ public void testAnonymousCanAlwaysLoadJARs() throws Exception {
     }
 
     @PresetData(DataSet.ANONYMOUS_READONLY)
-    public void testAnonymousCannotGetSecrets() throws Exception {
-        jenkins.setNodes(Collections.singletonList(createNewJnlpSlave("test")));
-        new WebClient().assertFails("computer/test/slave-agent.jnlp", HttpURLConnection.HTTP_FORBIDDEN);
+    @Test
+    public void anonymousCannotGetSecrets() throws Exception {
+        r.jenkins.setNodes(Collections.singletonList(createNewJnlpSlave("test")));
+        r.createWebClient().assertFails("computer/test/slave-agent.jnlp", HttpURLConnection.HTTP_FORBIDDEN);
     }
 
     @PresetData(DataSet.NO_ANONYMOUS_READACCESS)
     @SuppressWarnings("SleepWhileInLoop")
-    public void testServiceUsingDirectSecret() throws Exception {
+    @Test
+    public void serviceUsingDirectSecret() throws Exception {
         Slave slave = createNewJnlpSlave("test");
-        jenkins.setNodes(Collections.singletonList(slave));
-        new WebClient().goTo("computer/test/slave-agent.jnlp?encrypt=true", "application/octet-stream");
+        r.jenkins.setNodes(Collections.singletonList(slave));
+        r.createWebClient().goTo("computer/test/slave-agent.jnlp?encrypt=true", "application/octet-stream");
         String secret = slave.getComputer().getJnlpMac();
         // To watch it fail: secret = secret.replace('1', '2');
-        ProcessBuilder pb = new ProcessBuilder(JavaEnvUtils.getJreExecutable("java"), "-jar", Which.jarFile(Launcher.class).getAbsolutePath(), "-jnlpUrl", getURL() + "computer/test/slave-agent.jnlp", "-secret", secret);
-
-        pb = pb.inheritIO();
-
-        System.err.println("Running: " + pb.command());
-        Process p = pb.start();
+        File slaveJar = tmp.newFile();
+        FileUtils.copyURLToFile(new Slave.JnlpJar("slave.jar").getURL(), slaveJar);
+        Proc p = new hudson.Launcher.LocalLauncher(StreamTaskListener.fromStderr()).launch().
+            stdout(System.out).stderr(System.err).
+            cmds(JavaEnvUtils.getJreExecutable("java"), "-jar", slaveJar.getAbsolutePath(), "-jnlpUrl", r.getURL() + "computer/test/slave-agent.jnlp", "-secret", secret).
+            start();
         try {
-            for (int i = 0; i < /* one minute */600; i++) {
-                if (slave.getComputer().isOnline()) {
-                    System.err.println("JNLP slave successfully connected");
-                    Channel channel = slave.getComputer().getChannel();
```

Comment:
[diff for review](https://github.com/jenkinsci/jenkins/pull/3096/files?w=1#diff-139a67312141a030a142c3f3f69da215L115)